apps/labs/posts/narwhals-pycapsule.md

```diff
@@ -0,0 +1,199 @@
+between them!
+
+> **_NOTE:_** How many times can `agnostic_sum_i64_column` be called? Strictly
+> speaking, the PyCapsule Interface only guarantees that it can be called once for
+> any given input. In practice, however, all implementations that we're aware of
+> (except for DuckDB) allow for it to be called repeatedly. Discussion about the
+> DuckDB implementation is ongoing at https://github.com/duckdb/duckdb/discussions/15536.
+
+If you found the above example a little daunting, you may be wondering if a simpler
+solution exists which you can develop entirely in Python-land. Enter: Narwhals.
+
+## Python solution: Narwhals
+
+Narwhals is a lightweight, dependency-free, and extensible compatibility layer between
+dataframe libraries functions. To use it to write a dataframe-agnostic function,
+you need to follow three steps:
+
+- Call `narwhals.from_native` on the user's object.
+- Use the [Narwhals API](https://narwhals-dev.github.io/narwhals/api-reference/).
+- If you want to return to the user an object of the same kind that they started with,
+  call `to_native`.
+
+In this case, a Narwhals version of `agnostic_sum_i64_column` looks like this:
+
+```python
+import narwhals as nw
+from narwhals.typing import IntoFrame
+
+
+def agnostic_sum_i64_column_narwhals(df_native: IntoFrame, column_name: str) -> int:
+    lf = nw.from_native(df_native).lazy()
+    schema = lf.collect_schema()
+    if column_name not in schema:
+        msg = f"Column '{column_name}' not found, available columns are: {schema.names()}."
+        raise ValueError(msg)
+    if (dtype := schema[column_name]) != nw.Int64:
+        msg = f"Column '{column_name}' is of type {dtype}, expected Int64"
+        raise TypeError(msg)
+    df = lf.collect()
+    return df[column_name].sum()
+```
+
+Like above, we can now pass different inputs and get the same result:
+```python
+>>> agnostic_sum_i64_column_narwhals(df_pl, column_name="a")
+6
+>>> agnostic_sum_i64_column_narwhals(df_pd, column_name="a")
+6
+>>> agnostic_sum_i64_column_narwhals(df_pa, column_name="a")
+6
+>>> agnostic_sum_i64_column_narwhals(rel, column_name="a")
+6
+```
+
+So long as you stick to the Narwhals API, your code will keep working with all major dataframe libraries,
+as well as with new ones which may appear in the future!
+
+## Narwhals vs PyCapsule Interface: when to use one over the other, and when to use them together
+
+If the Narwhals API is extensive enough for your use-case, then this is arguably
+a simpler and easier solution than writing your own Rust function. On the other hand,
+if you write a Rust function using the PyCapsule Interface, then you have complete
+control over the data. So, when should you use which?
+
+Let's cover some scenarios:
+
+- If you want your dataframe logic to stay completely lazy where possible: use Narwhals.
+  The PyCapsule Interface requires you to materialise the data into memory immediately,
```

Comment:
If anything, I think it's the opposite; the Arrow Stream interface encourages laziness by being defined as a stream. It's only common in-memory implementations that fully materialize data.